README
nagios-plugins-uptime
=====================

This Nagios plugin checks how long the system has been running.

Usage

        check_uptime [--warning [@]start:end] [--critical [@]start:end]
        check_uptime --help
        check_uptime --version

Where

 * start <= end
 * start and ":" is not required if start=0
 * if range is of format "start:" and end is not specified, assume end is infinity
 * to specify negative infinity, use "~"
 * alert is raised if metric is outside start and end range (inclusive of endpoints)
 * if range starts with "@", then alert if inside this range (inclusive of endpoints)

Examples

        check_uptime
        check_uptime --warning 30: --critical 15:


## Source code

The source code can be also found at
https://sites.google.com/site/davidemadrisan/opensource


## Installation

This package uses GNU autotools for configuration and installation.

If you have cloned the git repository then you will need to run `autogen.sh`
or `autoreconf` (autoconf package) to generate the required files.

Run `./configure --help` to see a list of available install options.
The plugin will be installed by default into `LIBEXECDIR`.

It is highly likely that you will want to customise this location to suit your
needs, i.e.:

        ./configure --libexecdir=/usr/lib/nagios/plugins

After `./configure` has completed successfully run `make install` and you're
done!


## Supported Platforms

This package is written in plain C, making as few assumptions as possible, and
sticking closely to ANSI C/POSIX.

This is a list of platforms this Nagios plugin is known to compile and run on

        Linux with kernel 3.6 and glibc 2.16.0 (openmamba milestone 2.75.0)
        FreeBSD 8.2-RELEASE-p10
        AIX 6.1 with gcc 4.2.0
        Solaris 11.1 (SunOS solaris 5.11 11.1) with gcc 4.5.2
        OpenBSD 5.2 with gcc 4.2.1

Other Unixes that provide the POSIX function `clock_gettime` are probably supported.
Please drop me a line if you have found one.
Thanks to shellmix.com for providing a free FreeBSD shell account.


## Bugs

If you find a bug please create an issue in the project bug tracker at
https://github.com/madrisan/nagios-plugins-uptime/issues



NEWS
    Major changes between "nagios-plugins-uptime" versions 7 and 8

* Use the function 'clock_gettime' when available.
  This should make the plugin working on all the unlisted Unixes
  that provide this POSIX function.

======================================================================

    Major changes between "nagios-plugins-uptime" versions 6 and 7

* Fix build errors on OpenBSD.

======================================================================

    Major changes between "nagios-plugins-uptime" versions 5 and 6

* Rework help and version messages.
* COPYING: update to GPLv3. 
* Code refactoring.

======================================================================

    Major changes between "nagios-plugins-uptime" versions 4 and 5

* Add support for Solaris 11 x86 (SunOS solaris 5.11 11.1, gcc 4.5.2).

======================================================================

    Major changes between "nagios-plugins-uptime" versions 3 and 4

* Add support for AIX (6.1 with gcc 4.2.0).

======================================================================

    Major changes between "nagios-plugins-uptime" versions 2 and 3

* The plugin "nagios-plugins-linux-uptime" has been renamed to
  "nagios-plugins-uptime".
* Add support for FreeBSD (8.2-RELEASE-p10).
* Add perfdata to the plugin output.
* Update package documentation.

======================================================================

 Major changes between "nagios-plugins-linux-uptime" versions 1 and 2

* Autotoolize package build.
* Enable more compiler warnings when using gcc.

======================================================================


src/nputils.c
/*
 * License: GPL
 * Copyright (c) 2006 Nagios Plugins Development Team
 *
 * Library of useful functions for nagios plugins
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config.h"

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>

#include <string.h>
#ifdef HAVE_STRINGS_H
#include <strings.h>
#endif

#include "nputils.h"

int check_range (double, range *);
range *parse_range_string (char *);
void set_range_start (range *, double);
void set_range_end (range *, double);

/*
 * Returns TRUE if alert should be raised based on the range 
 */
int
check_range (double value, range * my_range)
{
  int no = FALSE;
  int yes = TRUE;

  if (my_range->alert_on == INSIDE)
    {
      no = TRUE;
      yes = FALSE;
    }

  if (my_range->end_infinity == FALSE && my_range->start_infinity == FALSE)
    {
      if ((my_range->start <= value) && (value <= my_range->end))
	return no;
      else
	return yes;
    }
  else if (my_range->start_infinity == FALSE
	   && my_range->end_infinity == TRUE)
    {
      if (my_range->start <= value)
	return no;
      else
	return yes;
    }
  else if (my_range->start_infinity == TRUE
	   && my_range->end_infinity == FALSE)
    {
      if (value <= my_range->end)
	return no;
      else
	return yes;
    }
  else
    return no;
}

int
get_status (double value, thresholds * my_thresholds)
{
  if (my_thresholds->critical != NULL)
    {
      if (check_range (value, my_thresholds->critical) == TRUE)
	return STATE_CRITICAL;
    }
  if (my_thresholds->warning != NULL)
    {
      if (check_range (value, my_thresholds->warning) == TRUE)
	return STATE_WARNING;
    }
  return STATE_OK;
}

void
set_range_start (range * this, double value)
{
  this->start = value;
  this->start_infinity = FALSE;
}

void
set_range_end (range * this, double value)
{
  this->end = value;
  this->end_infinity = FALSE;
}

range *
parse_range_string (char *str)
{
  range *temp_range;
  double start;
  double end;
  char *end_str;

  temp_range = (range *) malloc (sizeof (range));

  /*
   * Set defaults 
   */
  temp_range->start = 0;
  temp_range->start_infinity = FALSE;
  temp_range->end = 0;
  temp_range->end_infinity = TRUE;
  temp_range->alert_on = OUTSIDE;

  if (str[0] == '@')
    {
      temp_range->alert_on = INSIDE;
      str++;
    }

  end_str = index (str, ':');
  if (end_str != NULL)
    {
      if (str[0] == '~')
	temp_range->start_infinity = TRUE;
      else
	{
	  start = strtod (str, NULL);	/* Will stop at the ':' */
	  set_range_start (temp_range, start);
	}
      end_str++;		/* Move past the ':' */
    }
  else
    {
      end_str = str;
    }
  end = strtod (end_str, NULL);
  if (strcmp (end_str, "") != 0)
    set_range_end (temp_range, end);

  if (temp_range->start_infinity == TRUE ||
      temp_range->end_infinity == TRUE ||
      temp_range->start <= temp_range->end)
    {
      return temp_range;
    }
  free (temp_range);
  return NULL;
}

/*
 * returns 0 if okay, otherwise 1 
 */
int
set_thresholds (thresholds ** my_thresholds, char *warn_string,
		char *critical_string)
{
  thresholds *temp_thresholds = NULL;

  if ((temp_thresholds = malloc (sizeof (thresholds))) == NULL)
    {
      printf ("Cannot allocate memory: %s", strerror (errno));
      exit (STATE_UNKNOWN);
    }

  temp_thresholds->warning = NULL;
  temp_thresholds->critical = NULL;

  if (warn_string != NULL)
    {
      if ((temp_thresholds->warning =
	   parse_range_string (warn_string)) == NULL)
	{
	  return NP_RANGE_UNPARSEABLE;
	}
    }
  if (critical_string != NULL)
    {
      if ((temp_thresholds->critical =
	   parse_range_string (critical_string)) == NULL)
	{
	  return NP_RANGE_UNPARSEABLE;
	}
    }

  *my_thresholds = temp_thresholds;

  return 0;
}


src/getopt_long.c
/*	$NetBSD: getopt_long.c,v 1.10.2.1 2009/04/14 21:18:35 tron Exp $	*/

/*-
 * Copyright (c) 2000 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Dieter Baron and Thomas Klausner.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "config.h"

#include <assert.h>
#ifdef HAVE_ERR_H
#include <err.h>
#else
#define warnx printf
#endif
#include <errno.h>
#if defined(HAVE_GETOPT_H) && defined(HAVE_STRUCT_OPTION)
#include <getopt.h>
#else
#include "compat_getopt.h"
#endif
#include <stdlib.h>
#include <string.h>

#define REPLACE_GETOPT

#undef __UNCONST
#define __UNCONST(a)	((void *)(size_t)(const void *)(a))

#ifndef _DIAGASSERT
#define _DIAGASSERT assert
#endif

#ifdef REPLACE_GETOPT
#ifdef __weak_alias
__weak_alias(getopt,_getopt)
#endif
int	opterr = 1;		/* if error message should be printed */
int	optind = 1;		/* index into parent argv vector */
int	optopt = '?';		/* character checked for validity */
int	optreset;		/* reset getopt */
char    *optarg;		/* argument associated with option */
#elif HAVE_NBTOOL_CONFIG_H && !HAVE_DECL_OPTRESET
static int optreset;
#endif

#if 0
#ifdef __weak_alias
__weak_alias(getopt_long,_getopt_long)
#endif
#endif

#define IGNORE_FIRST	(*options == '-' || *options == '+')
#define PRINT_ERROR	((opterr) && ((*options != ':') \
				      || (IGNORE_FIRST && options[1] != ':')))
#define IS_POSIXLY_CORRECT (getenv("POSIXLY_CORRECT") != NULL)
#define PERMUTE         (!IS_POSIXLY_CORRECT && !IGNORE_FIRST)
/* XXX: GNU ignores PC if *options == '-' */
#define IN_ORDER        (!IS_POSIXLY_CORRECT && *options == '-')

/* return values */
#define	BADCH	(int)'?'
#define	BADARG		((IGNORE_FIRST && options[1] == ':') \
			 || (*options == ':') ? (int)':' : (int)'?')
#define INORDER (int)1

#define	EMSG	""

static int getopt_internal(int, char **, const char *);
static int gcd(int, int);
static void permute_args(int, int, int, char **);

static const char *place = EMSG; /* option letter processing */

/* XXX: set optreset to 1 rather than these two */
static int nonopt_start = -1; /* first non option argument (for permute) */
static int nonopt_end = -1;   /* first option after non options (for permute) */

/* Error messages */
static const char recargchar[] = "option requires an argument -- %c";
static const char recargstring[] = "option requires an argument -- %s";
static const char ambig[] = "ambiguous option -- %.*s";
static const char noarg[] = "option doesn't take an argument -- %.*s";
static const char illoptchar[] = "unknown option -- %c";
static const char illoptstring[] = "unknown option -- %s";


/*
 * Compute the greatest common divisor of a and b.
 */
static int
gcd(int a, int b)
{
	int c;

	c = a % b;
	while (c != 0) {
		a = b;
		b = c;
		c = a % b;
	}
	   
	return b;
}

/*
 * Exchange the block from nonopt_start to nonopt_end with the block
 * from nonopt_end to opt_end (keeping the same order of arguments
 * in each block).
 */
static void
permute_args(int panonopt_start, int panonopt_end, int opt_end, char **nargv)
{
	int cstart, cyclelen, i, j, ncycle, nnonopts, nopts, pos;
	char *swap;

	_DIAGASSERT(nargv != NULL);

	/*
	 * compute lengths of blocks and number and size of cycles
	 */
	nnonopts = panonopt_end - panonopt_start;
	nopts = opt_end - panonopt_end;
	ncycle = gcd(nnonopts, nopts);
	cyclelen = (opt_end - panonopt_start) / ncycle;

	for (i = 0; i < ncycle; i++) {
		cstart = panonopt_end+i;
		pos = cstart;
		for (j = 0; j < cyclelen; j++) {
			if (pos >= panonopt_end)
				pos -= nnonopts;
			else
				pos += nopts;
			swap = nargv[pos];
			nargv[pos] = nargv[cstart];
			nargv[cstart] = swap;
		}
	}
}

/*
 * getopt_internal --
 *	Parse argc/argv argument vector.  Called by user level routines.
 *  Returns -2 if -- is found (can be long option or end of options marker).
 */
static int
getopt_internal(int nargc, char **nargv, const char *options)
{
	char *oli;				/* option letter list index */
	int optchar;

	_DIAGASSERT(nargv != NULL);
	_DIAGASSERT(options != NULL);

	optarg = NULL;

	/*
	 * XXX Some programs (like rsyncd) expect to be able to
	 * XXX re-initialize optind to 0 and have getopt_long(3)
	 * XXX properly function again.  Work around this braindamage.
	 */
	if (optind == 0)
		optind = 1;

	if (optreset)
		nonopt_start = nonopt_end = -1;
start:
	if (optreset || !*place) {		/* update scanning pointer */
		optreset = 0;
		if (optind >= nargc) {          /* end of argument vector */
			place = EMSG;
			if (nonopt_end != -1) {
				/* do permutation, if we have to */
				permute_args(nonopt_start, nonopt_end,
				    optind, nargv);
				optind -= nonopt_end - nonopt_start;
			}
			else if (nonopt_start != -1) {
				/*
				 * If we skipped non-options, set optind
				 * to the first of them.
				 */
				optind = nonopt_start;
			}
			nonopt_start = nonopt_end = -1;
			return -1;
		}
		if ((*(place = nargv[optind]) != '-')
		    || (place[1] == '\0')) {    /* found non-option */
			place = EMSG;
			if (IN_ORDER) {
				/*
				 * GNU extension: 
				 * return non-option as argument to option 1
				 */
				optarg = nargv[optind++];
				return INORDER;
			}
			if (!PERMUTE) {
				/*
				 * if no permutation wanted, stop parsing
				 * at first non-option
				 */
				return -1;
			}
			/* do permutation */
			if (nonopt_start == -1)
				nonopt_start = optind;
			else if (nonopt_end != -1) {
				permute_args(nonopt_start, nonopt_end,
				    optind, nargv);
				nonopt_start = optind -
				    (nonopt_end - nonopt_start);
				nonopt_end = -1;
			}
			optind++;
			/* process next argument */
			goto start;
		}
		if (nonopt_start != -1 && nonopt_end == -1)
			nonopt_end = optind;
		if (place[1] && *++place == '-') {	/* found "--" */
			place++;
			return -2;
		}
	}
	if ((optchar = (int)*place++) == (int)':' ||
	    (oli = strchr(options + (IGNORE_FIRST ? 1 : 0), optchar)) == NULL) {
		/* option letter unknown or ':' */
		if (!*place)
			++optind;
		if (PRINT_ERROR)
			warnx(illoptchar, optchar);
		optopt = optchar;
		return BADCH;
	}
	if (optchar == 'W' && oli[1] == ';') {		/* -W long-option */
		/* XXX: what if no long options provided (called by getopt)? */
		if (*place) 
			return -2;

		if (++optind >= nargc) {	/* no arg */
			place = EMSG;
			if (PRINT_ERROR)
				warnx(recargchar, optchar);
			optopt = optchar;
			return BADARG;
		} else				/* white space */
			place = nargv[optind];
		/*
		 * Handle -W arg the same as --arg (which causes getopt to
		 * stop parsing).
		 */
		return -2;
	}
	if (*++oli != ':') {			/* doesn't take argument */
		if (!*place)
			++optind;
	} else {				/* takes (optional) argument */
		optarg = NULL;
		if (*place)			/* no white space */
			optarg = __UNCONST(place);
		/* XXX: disable test for :: if PC? (GNU doesn't) */
		else if (oli[1] != ':') {	/* arg not optional */
			if (++optind >= nargc) {	/* no arg */
				place = EMSG;
				if (PRINT_ERROR)
					warnx(recargchar, optchar);
				optopt = optchar;
				return BADARG;
			} else
				optarg = nargv[optind];
		}
		place = EMSG;
		++optind;
	}
	/* dump back option letter */
	return optchar;
}

#ifdef REPLACE_GETOPT
/*
 * getopt --
 *	Parse argc/argv argument vector.
 *
 * [eventually this will replace the real getopt]
 */
int
getopt(nargc, nargv, options)
	int nargc;
	char * const *nargv;
	const char *options;
{
	int retval;

	_DIAGASSERT(nargv != NULL);
	_DIAGASSERT(options != NULL);

	retval = getopt_internal(nargc, __UNCONST(nargv), options);
	if (retval == -2) {
		++optind;
		/*
		 * We found an option (--), so if we skipped non-options,
		 * we have to permute.
		 */
		if (nonopt_end != -1) {
			permute_args(nonopt_start, nonopt_end, optind,
				       __UNCONST(nargv));
			optind -= nonopt_end - nonopt_start;
		}
		nonopt_start = nonopt_end = -1;
		retval = -1;
	}
	return retval;
}
#endif

/*
 * getopt_long --
 *	Parse argc/argv argument vector.
 */
int
getopt_long(int nargc, char * const *nargv, const char *options,
    const struct option *long_options, int *idx)
{
	int retval;

#define IDENTICAL_INTERPRETATION(_x, _y)				\
	(long_options[(_x)].has_arg == long_options[(_y)].has_arg &&	\
	 long_options[(_x)].flag == long_options[(_y)].flag &&		\
	 long_options[(_x)].val == long_options[(_y)].val)

	_DIAGASSERT(nargv != NULL);
	_DIAGASSERT(options != NULL);
	_DIAGASSERT(long_options != NULL);
	/* idx may be NULL */

	retval = getopt_internal(nargc, __UNCONST(nargv), options);
	if (retval == -2) {
		char *current_argv, *has_equal;
		size_t current_argv_len;
		int i, ambiguous, match;

		current_argv = __UNCONST(place);
		match = -1;
		ambiguous = 0;

		optind++;
		place = EMSG;

		if (*current_argv == '\0') {		/* found "--" */
			/*
			 * We found an option (--), so if we skipped
			 * non-options, we have to permute.
			 */
			if (nonopt_end != -1) {
				permute_args(nonopt_start, nonopt_end,
				    optind, __UNCONST(nargv));
				optind -= nonopt_end - nonopt_start;
			}
			nonopt_start = nonopt_end = -1;
			return -1;
		}
		if ((has_equal = strchr(current_argv, '=')) != NULL) {
			/* argument found (--option=arg) */
			current_argv_len = has_equal - current_argv;
			has_equal++;
		} else
			current_argv_len = strlen(current_argv);
	    
		for (i = 0; long_options[i].name; i++) {
			/* find matching long option */
			if (strncmp(current_argv, long_options[i].name,
			    current_argv_len))
				continue;

			if (strlen(long_options[i].name) ==
			    (unsigned)current_argv_len) {
				/* exact match */
				match = i;
				ambiguous = 0;
				break;
			}
			if (match == -1)		/* partial match */
				match = i;
			else if (!IDENTICAL_INTERPRETATION(i, match))
				ambiguous = 1;
		}
		if (ambiguous) {
			/* ambiguous abbreviation */
			if (PRINT_ERROR)
				warnx(ambig, (int)current_argv_len,
				     current_argv);
			optopt = 0;
			return BADCH;
		}
		if (match != -1) {			/* option found */
		        if (long_options[match].has_arg == no_argument
			    && has_equal) {
				if (PRINT_ERROR)
					warnx(noarg, (int)current_argv_len,
					     current_argv);
				/*
				 * XXX: GNU sets optopt to val regardless of
				 * flag
				 */
				if (long_options[match].flag == NULL)
					optopt = long_options[match].val;
				else
					optopt = 0;
				return BADARG;
			}
			if (long_options[match].has_arg == required_argument ||
			    long_options[match].has_arg == optional_argument) {
				if (has_equal)
					optarg = has_equal;
				else if (long_options[match].has_arg ==
				    required_argument) {
					/*
					 * optional argument doesn't use
					 * next nargv
					 */
					optarg = nargv[optind++];
				}
			}
			if ((long_options[match].has_arg == required_argument)
			    && (optarg == NULL)) {
				/*
				 * Missing argument; leading ':'
				 * indicates no error should be generated
				 */
				if (PRINT_ERROR)
					warnx(recargstring, current_argv);
				/*
				 * XXX: GNU sets optopt to val regardless
				 * of flag
				 */
				if (long_options[match].flag == NULL)
					optopt = long_options[match].val;
				else
					optopt = 0;
				--optind;
				return BADARG;
			}
		} else {			/* unknown option */
			if (PRINT_ERROR)
				warnx(illoptstring, current_argv);
			optopt = 0;
			return BADCH;
		}
		if (long_options[match].flag) {
			*long_options[match].flag = long_options[match].val;
			retval = 0;
		} else 
			retval = long_options[match].val;
		if (idx)
			*idx = match;
	}
	return retval;
#undef IDENTICAL_INTERPRETATION
}


src/check_uptime.c
/*
 * License: GPL
 * Copyright (c) 2010,2012,2013 Davide Madrisan <davide.madrisan@gmail.com>
 *
 * A Nagios plugin to check how long the system has been running
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config.h"

#include <assert.h>
#include <fcntl.h>
#if HAVE_GETOPT_H
#include <getopt.h>
#else
#include <compat_getopt.h>
#endif

#if HAVE_KSTAT_H
#include <kstat.h>
#endif

#if HAVE_LIBPERFSTAT
#include <sys/protosw.h>
#include <libperfstat.h>
#endif

#include <stdio.h>
#include <stdlib.h>

#if HAVE_SYS_SYSINFO_H
#include <sys/sysinfo.h>
#endif

#if TIME_WITH_SYS_TIME
#include <sys/time.h>
#include <time.h>
#else
#if HAVE_SYS_TIME_H
#include <sys/time.h>
#else
#include <time.h>
#endif
#endif

#include <unistd.h>

#ifdef HAVE_SYS_PARAM_H
#include <sys/param.h>
#endif

#if HAVE_SYS_SYSCTL_H
#include <sys/sysctl.h>
#endif

#ifdef HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif

#include "nputils.h"

static const char *program_name = "check_update";
static const char *program_version = PACKAGE_VERSION;
static const char *program_copyright =
  "Copyright (C) 2010,2012-2013 Davide Madrisan <" PACKAGE_BUGREPORT ">";

#define BUFSIZE 127
static char buf[BUFSIZE + 1];
static char result_line[BUFSIZE + 1], perfdata_line[BUFSIZE + 1];

double uptime (void);
char *sprint_uptime (double);

static void __attribute__ ((__noreturn__)) print_version (void)
{
  printf ("%s, version %s\n", program_name, program_version);
  printf ("%s\n", program_copyright);
  fputs ("\
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n\n\
This is free software; you are free to change and redistribute it.\n\
There is NO WARRANTY, to the extent permitted by law.\n", stdout);

  exit (STATE_OK);
}

static struct option const longopts[] = {
  {(char *) "critical", required_argument, NULL, 'c'},
  {(char *) "warning", required_argument, NULL, 'w'},
  {(char *) "help", no_argument, NULL, 'h'},
  {(char *) "version", no_argument, NULL, 'V'},
  {NULL, 0, NULL, 0}
};

static void __attribute__ ((__noreturn__)) usage (FILE * out)
{
  fprintf (out,
	   "%s, version %s - check how long the system has been running.\n",
	   program_name, program_version);
  fprintf (out, "%s\n\n", program_copyright);
  fprintf (out, "Usage: %s [OPTION]\n\n", program_name);
  fputs ("\
Options:\n\
  -w, --warning [@]start:end]   warning threshold\n\
  -c, --critical [@]start:end]   critical threshold\n\
  -h, --help            display this help and exit\n\
  -v, --version         output version information and exit\n\n", out);

  fputs ("\
Where:\n\
  1. start <= end\n\
  2. start and \":\" is not required if start=0\n\
  3. if range is of format \"start:\" and end is not specified, assume end is infinity\n\
  4. to specify negative infinity, use \"~\"\n\
  5. alert is raised if metric is outside start and end range (inclusive of endpoints)\n\
  6. if range starts with \"@\", then alert if inside this range (inclusive of endpoints)\n\n", out);
  fprintf (out, "Examples:\n  %s\n  %s --warning 30: --critical 15:\n",
	   program_name, program_name);

  exit (out == stderr ? STATE_UNKNOWN : STATE_OK);
}

/* assume uptime never be zero seconds in practice */
#define UPTIME_RET_FAIL  0

double
uptime ()
{
#if defined(HAVE_STRUCT_SYSINFO_WITH_UPTIME)	/* Linux */

  struct sysinfo info;

  if (0 != sysinfo (&info))
    {
      perror ("cannot get the system uptime");
      return UPTIME_RET_FAIL;
    }

  return info.uptime;

#elif defined(HAVE_FUNCTION_SYSCTL_KERN_BOOTTIME)	/* FreeBSD */

  int mib[] = { CTL_KERN, KERN_BOOTTIME };
  struct timeval system_uptime;
  size_t len = sizeof (system_uptime);

  if (0 != sysctl (mib, 2, &system_uptime, &len, NULL, 0))
    return UPTIME_RET_FAIL;

  return (time (NULL) - system_uptime.tv_sec);

#elif defined(HAVE_KSTAT_H)	/* Solaris */

  kstat_ctl_t *kc;
  kstat_t *ksp;
  kstat_named_t *knp;

  time_t now;

  if (NULL == (kc = kstat_open ()))
    return UPTIME_RET_FAIL;

  if (NULL !=
      (ksp = kstat_lookup (kc, (char *) "unix", 0, (char *) "system_misc")))
    {
      if (-1 != kstat_read (kc, ksp, 0))
	{
	  if (NULL != (knp = kstat_data_lookup (ksp, (char *) "boot_time")))
	    {
	      time (&now);
	      kstat_close (kc);
	      return (difftime (now, (time_t) knp->value.ul));
	    }
	}
    }

  kstat_close (kc);
  return UPTIME_RET_FAIL;

#elif defined(HAVE_LIBPERFSTAT)	/* AIX */

  long herz = 0;
  perfstat_cpu_total_t ps_cpu_total;

  herz = sysconf (_SC_CLK_TCK);
  /* make sure we do not divide by 0 */
  assert (herz);

  if (-1 ==
      perfstat_cpu_total (NULL, &ps_cpu_total, sizeof (ps_cpu_total), 1))
    return UPTIME_RET_FAIL;

  return ((double) ps_cpu_total.lbolt / herz);

#elif defined(HAVE_CLOCK_GETTIME_MONOTONIC) /* POSIX.1-2001 */

  struct timespec t;

  clock_gettime (CLOCK_MONOTONIC, &t);
  if (t.tv_sec > 0)
    return t.tv_sec;
  else
    return UPTIME_RET_FAIL;

#else

  return UPTIME_RET_FAIL;

#endif
}

char *
sprint_uptime (double uptime_secs)
{
  int upminutes, uphours, updays;
  int pos = 0;

  updays = (int) uptime_secs / (60 * 60 * 24);
  if (updays)
    pos +=
      snprintf (buf, BUFSIZE, "%d day%s ", updays, (updays != 1) ? "s" : "");
  upminutes = (int) uptime_secs / 60;
  uphours = upminutes / 60;
  uphours = uphours % 24;
  upminutes = upminutes % 60;

  if (uphours)
    {
      pos +=
	snprintf (buf + pos, BUFSIZE - pos, "%d hour%s %d min", uphours,
		  (uphours != 1) ? "s" : "", upminutes);
    }
  else
    pos += snprintf (buf + pos, BUFSIZE - pos, "%d min", upminutes);

  return buf;
}

int
main (int argc, char **argv)
{
  int c, uptime_mins, status;
  char *critical = NULL, *warning = NULL;
  double uptime_secs;
  thresholds *my_threshold = NULL;

  while ((c = getopt_long (argc, argv, "c:w:hV", longopts, NULL)) != -1)
    {
      switch (c)
	{
	default:
	  usage (stderr);
	  break;
	case 'c':
	  critical = optarg;
	  break;
	case 'w':
	  warning = optarg;
	  break;
	case 'h':
	  usage (stdout);
	  break;
	case 'V':
	  print_version ();
	  break;
	}
    }

  status = set_thresholds (&my_threshold, warning, critical);
  if (status == NP_RANGE_UNPARSEABLE)
    usage (stderr);

  if (UPTIME_RET_FAIL != (uptime_secs = uptime ()))
    {
      uptime_mins = (int) uptime_secs / 60;
      status = get_status (uptime_mins, my_threshold);
      free (my_threshold);

      switch (status)
	{
	case STATE_CRITICAL:
	  c = snprintf (result_line, BUFSIZE, "UPTIME CRITICAL:");
	  break;
	case STATE_WARNING:
	  c = snprintf (result_line, BUFSIZE, "UPTIME WARNING:");
	  break;
	case STATE_OK:
	  c = snprintf (result_line, BUFSIZE, "UPTIME OK:");
	  break;
	}

      snprintf (result_line + c, BUFSIZE - c, " %s",
		sprint_uptime (uptime_secs));
      snprintf (perfdata_line, BUFSIZE, "uptime=%d", uptime_mins);

      printf ("%s|%s\n", result_line, perfdata_line);
    }
  else
    {
      c = snprintf (result_line, BUFSIZE,
		    "UPTIME UNKNOWN: can't get system uptime counter");
      status = STATE_UNKNOWN;
    }

  return status;
}
